### PR TITLE
schedule: combine store limit and store balance rate

### LINKED ISSUE #2245: UCP: Combine the store balance rate and store limit
## Description

Currently, we have a config item named `store-balance-rate` to control the balance speed of a store. This item will initialize the store limit when the server starts or restarts. We use `limit [<store_id>|<all> <rate>]` in pd-ctl to set the store limit when the pd is running and the setting state is not persistent. When the pd is restarted, the value will back to the value of `store-balance-rate`. 
Since we need to use two ways to control the store limit, it might be confused sometimes and not friendly enough. These two ways can be combined.

## Score

- 200

## Mentor

@rleungx 

## Recommended Skills

- Golang

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve? 
Close #2245.

### What is changed and how it works?
This PR mainly does these things:
- remove store-balance-rate
- persist store limit in schedule config
- change the auto store limit policy
- rename `*region` to `*peer`
- move `RemoveScheduler` from `Options` to `Cluster` interface
- some updates for testing and doc.

### Check List 

Tests 

 - Unit test

### Release note 
- No release note

### CODE REVIEW

File: server/api/store.go [hunk 1]
```diff
@@ -609,7 +586,7 @@ func (filter *storeStateFilter) filter(stores []*metapb.Store) []*metapb.Store {
 
 func getStoreLimitType(input map[string]interface{}) (storelimit.Type, error) {
 	typeNameIface, ok := input["type"]
-	typeValue := storelimit.RegionAdd
+	typeValue := storelimit.AddPeer
```
Comment (lhy1024) [inline server/api/store.go:589]: Will we consider to `RemovePeer`?

File: server/schedule/operator_controller_test.go [hunk 2]
```diff
@@ -543,17 +544,18 @@ func (t *testOperatorControllerSuite) TestStoreLimitWithMerge(c *C) {
 	}
 
 	tc.AddLeaderStore(1, 10)
+	tc.AddLeaderStore(2, 10)
+	tc.AddLeaderStore(3, 10)
 	tc.AddLeaderStore(4, 10)
 	tc.AddLeaderStore(5, 10)
+	tc.AddLeaderStore(6, 10)
```
Comment (lhy1024) [inline server/schedule/operator_controller_test.go:551]: how about replacing with a loop

File: server/handler.go [hunk 3]
```diff
@@ -418,32 +418,56 @@ func (h *Handler) GetHistory(start time.Time) ([]operator.OpHistory, error) {
...
+		}
+	case storelimit.RemovePeer:
+		for storeID := range cfg.StoreLimit {
+			sc := config.StoreLimitConfig{
+				AddPeer:    cfg.StoreLimit[storeID].AddPeer,
+				RemovePeer: ratePerMin,
+			}
+			cfg.StoreLimit[storeID] = sc
+		}
 	}
```
Comment (lhy1024) [inline server/handler.go:440]:
```suggestion
for storeID := range cfg.StoreLimit {
	cfg.StoreLimit[storeID][limitType] = ratePerMin 
}
```

Comment (lhy1024) [reply server/handler.go:440]: Are you worried that there may be `type` not initialized?

File: server/cluster/cluster.go [hunk 4]
```diff
@@ -1673,6 +1668,47 @@ func (c *RaftCluster) GetStoreLimiter() *StoreLimiter {
...
+	sc := config.StoreLimitConfig{
+		AddPeer:    defaultStoreLimit,
+		RemovePeer: defaultStoreLimit,
+	}
+	cfg.StoreLimit[storeID] = sc
+	c.opt.SetScheduleConfig(cfg)
+}
+
+// RemoveStoreLimit remove a store limit for a given store ID.
+func (c *RaftCluster) RemoveStoreLimit(storeID uint64) {
```
Comment (lhy1024) [inline server/cluster/cluster.go:1693]: how about setting them with default value?

Comment (lhy1024) [inline server/cluster/cluster.go:1693]: how about setting them default value?

File: server/api/store.go (hunk 1 above)
Comment (rleungx) [reply server/api/store.go:589]: The default output is `AddPeer`. It is the same as before.

File: tools/pd-ctl/pdctl/command/config_command.go [hunk 5]
```diff
@@ -217,6 +217,7 @@ func showConfigCommandFunc(cmd *cobra.Command, args []string) {
 
 	delete(scheduleConfig, "schedulers-v2")
 	delete(scheduleConfig, "schedulers-payload")
+	delete(scheduleConfig, "store-limit")
```
[the remark this answers is not in the document]
Comment (rleungx) [reply tools/pd-ctl/pdctl/command/config_command.go:220]: When there are too many stores, the output will be noisy.

File: pkg/mock/mockcluster/mockcluster.go [hunk 6]
```diff
@@ -539,6 +546,11 @@ func (mc *Cluster) GetMaxReplicas() int {
 	return mc.ScheduleOptions.GetMaxReplicas()
 }
 
+// GetStoreLimitByType mocks method.
```
[the remark this answers is not in the document]
Comment (rleungx) [reply pkg/mock/mockcluster/mockcluster.go:549]: what do you mean "emphasize ByType"?

File: server/handler.go (hunk 3 above)
Comment (rleungx) [reply server/handler.go:440]: `StoreLimit[storeID]` is a struct not a map.

File: server/cluster/cluster.go (hunk 4 above)
Comment (rleungx) [reply server/cluster/cluster.go:1693]: Can you explain more?

File: pkg/mock/mockcluster/mockcluster.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (lhy1024) [reply pkg/mock/mockcluster/mockcluster.go:549]: how about replacing with `GetStoreLimit`？

File: server/cluster/cluster.go (hunk 4 above)
Comment (lhy1024) [reply server/cluster/cluster.go:1693]: When we have removed the store limit, is the store unlimited?

Comment (rleungx) [reply server/cluster/cluster.go:1693]: The offline process is handled by `SetStoreLimit`

File: pkg/mock/mockcluster/mockcluster.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (rleungx) [reply pkg/mock/mockcluster/mockcluster.go:549]: I think both `GetStoreLimit` and `GetStoreLimitByType` are ok here.

File: server/config/config.go [hunk 7]
```diff
@@ -534,8 +534,8 @@ type ScheduleConfig struct {
 	// If the number of times a region hits the hot cache is greater than this
 	// threshold, it is considered a hot region.
 	HotRegionCacheHitsThreshold uint64 `toml:"hot-region-cache-hits-threshold" json:"hot-region-cache-hits-threshold"`
-	// StoreBalanceRate is the maximum of balance rate for each store.
-	StoreBalanceRate float64 `toml:"store-balance-rate" json:"store-balance-rate"`
+	// StoreLimit is the limit of scheduling for stores.
+	StoreLimit map[uint64]StoreLimitConfig `toml:"store-limit" json:"store-limit"`
```
Comment (nolouch) [inline server/config/config.go:538]: Can we keep this first? Need to ensure API compatibility

File: tools/pd-ctl/README.md [hunk 8]
```diff
@@ -645,16 +644,16 @@ Usage:
 >> store label 1 zone cn               // Set the value of the label with the "zone" key to "cn" for the store with the store id of 1
 >> store weight 1 5 10                 // Set the leader weight to 5 and region weight to 10 for the store with the store id of 1
 >> store remove-tombstone              // Remove stores that are in tombstone state
->> store limit                         // Show limits of adding region operation for all stores
->> store limit region-add              // Show limits of adding region operation for all stores
->> store limit region-remove           // Show limits of removing region operation for all stores
+>> store limit                         // Show limits of adding peer operation for all stores
+>> store limit add-peer                // Show limits of adding peer operation for all stores
```
Comment (nolouch) [inline tools/pd-ctl/README.md:648]: we need a warning like:
```
`region-add` is deprecated, please use `add-peer`
```
if the user uses old command.
and I more likely use old command also works.

File: server/statistics/store_collection.go [hunk 9]
```diff
@@ -148,7 +148,6 @@ func (s *storeStatistics) Collect() {
 	configs["high-space-ratio"] = s.opt.GetHighSpaceRatio()
 	configs["low-space-ratio"] = s.opt.GetLowSpaceRatio()
 	configs["tolerant-size-ratio"] = s.opt.GetTolerantSizeRatio()
-	configs["store-balance-rate"] = s.opt.GetStoreBalanceRate()
```
[the remark this answers is not in the document]
Comment (rleungx) [reply server/statistics/store_collection.go:151]: It will show the message now.

File: server/config/config.go [hunk 10]
```diff
@@ -535,7 +574,10 @@ type ScheduleConfig struct {
 	// threshold, it is considered a hot region.
 	HotRegionCacheHitsThreshold uint64 `toml:"hot-region-cache-hits-threshold" json:"hot-region-cache-hits-threshold"`
 	// StoreBalanceRate is the maximum of balance rate for each store.
-	StoreBalanceRate float64 `toml:"store-balance-rate" json:"store-balance-rate"`
+	// WARN: StoreBalanceRate is deprecated.
+	StoreBalanceRate float64 `toml:"store-balance-rate" json:"store-balance-rate,omitempty"`
+	// StoreLimit is the limit of scheduling for stores.
+	StoreLimit map[uint64]StoreLimitConfig `toml:"store-limit" json:"store-limit"`
```
Comment (nolouch) [inline server/config/config.go:580]: `config show` may show too many entries?

File: tools/pd-ctl/README.md [hunk 11]
```diff
@@ -645,16 +644,16 @@ Usage:
...
+>> store limit remove-peer             // Show limits of removing peer operation for all stores
 >> store limit all 5                   // Limit 5 adding region operations per minute for all stores
```
Comment (nolouch) [inline tools/pd-ctl/README.md:650]:
```
store limit all 5
```
the command only set for `add-peer`,  can we set `remove-peer` at the same time?

File: server/config/config.go (hunk 10 above)
Comment (rleungx) [reply server/config/config.go:580]: `config show` will not but `config show all` will

File: tools/pd-ctl/README.md (hunk 11 above)
Comment (rleungx) [reply tools/pd-ctl/README.md:650]: I'm not sure it's better. What is your opinion? @lhy1024

Comment (nolouch) [reply tools/pd-ctl/README.md:650]: Before, the user only needed one command. Now users think that the speed has been released, but in fact there is still a speed limit for `remove-peer`.

**Comment by rleungx** [comment]
@lhy1024 @nolouch PTAL

File: server/handler.go [hunk 12]
```diff
@@ -418,32 +418,58 @@ func (h *Handler) GetHistory(start time.Time) ([]operator.OpHistory, error) {
 }
 
 // SetAllStoresLimit is used to set limit of all stores.
-func (h *Handler) SetAllStoresLimit(rate float64, limitType storelimit.Type) error {
-	c, err := h.GetOperatorController()
-	if err != nil {
-		return err
+func (h *Handler) SetAllStoresLimit(ratePerMin float64, limitType storelimit.Type) error {
```
Comment (lhy1024) [inline server/handler.go:421]: I found there is a lot of code about set `storelimit`, I don’t know if they can be simplified

**Pull request merged**